sharding/types.go

```diff
@@ -20,6 +20,10 @@
 
 package sharding
 
+
+// HashGen generates HashFn
+type HashGen func(l uint32) HashFn
```

Comment:
HashGen sounds a little ambiguous.  Perhaps call this `NewHashShardRangeFn` e.g.:
```go
type NewHashShardRangeFn(length uint32) HashFn
```